### DISCUSSION TITLE
(NextJS App Router) How to pass cookies when server is prerendering a client component?

### DISCUSSION THREAD

**Discussion by knthmn** [trpc/trpc#6095] [Q&A]
I am following the [guide](https://trpc.io/docs/client/react/server-components#6-create-a-trpc-caller-for-server-components) for setting up TRPC with NextJS App Router. I have the following client component

```tsx
"use client";

import { trpc } from "@/trpc/client";
import { useEffect } from "react";

export function HelloClientComponent() {
  const [response] = trpc.hello.useSuspenseQuery({ text: "knthmn" });
  return <div>{response.greeting}</div>;  // greeting is "hello, your userId is {userId}"
}
``` 
During sever rendering of the client component, the cookies do not get passed on and the message is rendered with userId = undefined. The rendered HTML and JS is sent to the client, and the client hydrates and calls the TRPC route again with proper cookies resulting in a hydration mismatch.

Is the setup wrong? How do I ensure that the cookies get passed on properly when server is rendering a client component?
Reproduction example: https://github.com/knthmn/trpc-rsc-cookies-example

Cheers,
knthmn

**Comment by nrathi** [comment]
What did you end up doing?

**Reply by knthmn** [reply]
I ended up using this https://trpc.io/docs/client/headers, if window does not exist, `headers()` is dynamically imported and called

**Reply by Icestonks** [reply]
did you try prefetching in your server component?

**Reply by knthmn** [reply]
Hi all,

I do not need to prefetch, see the changes I made here where I dynamically import `cookies()`.
https://github.com/knthmn/trpc-rsc-cookies-example/commit/bbb902cfa6ac0606d482cb9c8fd80da874860635
I think this code should be safe, since the headers are dynamically resolved every time so it does not leak sessions between users.
